### PR TITLE
[GOBBLIN-1764] Emit observability event

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1764

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

GobblinTrackingEvents are not suitable for observability platforms at scale due to their frequency of emission and lack of standardized schema.

GaaSObservabilityEvent is a new event type that provides a job summary from pipelines in GaaS. It differs from GobblinTrackingEvents as it runs once per job pipeline, and it designed to be easily queryable and alert on. It will have fixed schema, unlike GobblinTrackingEvents, which puts a map of job information into the `metadata` field which may or may not exist and is dependent on the execution framework to provide. This makes GTE more frail and difficult for users and other systems to reliably read from this event.

We want to emit this observability event from GaaS by deriving it from a job's job status. Since this feature is Experimental and WIP, it is not expected to fill out all of the fields immediately.

This creates a template for a configurable GaaSObservabilityEventProducer, and the emission logic can be implemented by any derived class that will define the message sending mechanism. 

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 1]
```diff
@@ -266,6 +277,11 @@ static void addJobStatusToStateStore(org.apache.gobblin.configuration.State jobS
 
       modifyStateIfRetryRequired(jobStatus);
       stateStore.put(storeName, tableName, jobStatus);
+
+      if (isStateTransitionToFinal(jobStatus, states) && eventProducer.isPresent()) {
+        log.info("I am here");
```
Comment (homatthew) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:282]: Did you mean to leave this here 😆

Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:282]: Oops I was debugging something and it got caught 😂

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 2]
```diff
@@ -219,7 +229,8 @@ protected void processMessage(DecodeableKafkaRecord<byte[],byte[]> message) {
    * @throws IOException
    */
   @VisibleForTesting
-  static void addJobStatusToStateStore(org.apache.gobblin.configuration.State jobStatus, StateStore stateStore)
+  static void addJobStatusToStateStore(org.apache.gobblin.configuration.State jobStatus, StateStore stateStore,
+      Optional<GaaSObservabilityEventProducer> eventProducer)
```
Comment (homatthew) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:233]: Curious on our philosophy  for using optional as a parameter. Why not use nulls + overloading. Java optionals are best used as a return value but as an input, I think it makes the developer experience less readable (There are plenty of debates on it on the internet). 

Although, I have seen the pattern used in some other parts of Gobblin, so clearly past developers felt it was okay.

Comment (phet) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:233]: I've not heard that debate....

overall, the idea w/ `Optional` is to encode possible absence in the type system where it can be checked statically and automatically.  `null` is notoriously furtive... take for instance this very impl. (here), which does not check `stateStore != null`--but should it?  the practice I follow is that when that ought to be checked, the expectation should be stated, not in javadoc or other code comments, but via `Optional`.

again, I've not heard specific concern w/ use as a parameter type.  quite the opposite, in fact: the use is noted to bring clear benefit.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/troubleshooter/JobIssueEventHandler.java [hunk 3]
```diff
@@ -50,6 +51,7 @@ public class JobIssueEventHandler {
   private static final Logger issueLogger =
       LoggerFactory.getLogger("org.apache.gobblin.runtime.troubleshooter.JobIssueLogger");
 
+  @Getter
```
Comment (phet) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/troubleshooter/JobIssueEventHandler.java:54]: any expectations to document here, such as only read-only use... or doesn't matter?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaAvroJobStatusMonitor.java [hunk 4]
```diff
@@ -65,9 +65,9 @@ public class KafkaAvroJobStatusMonitor extends KafkaJobStatusMonitor {
   private Meter messageParseFailures;
 
   public KafkaAvroJobStatusMonitor(String topic, Config config, int numThreads,
-      JobIssueEventHandler jobIssueEventHandler)
+      JobIssueEventHandler jobIssueEventHandler, boolean instrumentationEnabled)
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaAvroJobStatusMonitor.java:68]: instrumentation is vague, but even there suggests something different, such as us measuring how fast the class operates as it handles messages.  why not call it what it is--`shouldEmitObervabilityEvents`?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 5]
```diff
@@ -288,6 +303,15 @@ private static void modifyStateIfRetryRequired(org.apache.gobblin.configuration.
     state.removeProp(TimingEvent.FlowEventConstants.DOES_CANCELED_FLOW_MERIT_RETRY);
   }
 
+  private static boolean isStateTransitionToFinal(org.apache.gobblin.configuration.State currentState, List<org.apache.gobblin.configuration.State> prevStates) {
+    Set<String> finalStates = ImmutableSet.of(ExecutionStatus.COMPLETE.name(), ExecutionStatus.CANCELLED.name(), ExecutionStatus.FAILED.name());
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:307]: may be better as a class `static`

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 6]
```diff
@@ -288,6 +303,15 @@ private static void modifyStateIfRetryRequired(org.apache.gobblin.configuration.
...
+    if (prevStates.size() == 0) {
+      return finalStates.contains(currentState.getProp(JobStatusRetriever.EVENT_NAME_FIELD));
+    }
+    return currentState.contains(JobStatusRetriever.EVENT_NAME_FIELD) && finalStates.contains(currentState.getProp(JobStatusRetriever.EVENT_NAME_FIELD))
+        && !finalStates.contains(prevStates.get(prevStates.size()-1).getProp(JobStatusRetriever.EVENT_NAME_FIELD));
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:312]: thought marginally more processing, it more clearly captures the intent and handles a corner case (if even possible) of the state flip flopping from final to non-final back to final:
```
Stream<Boolean> statesFinality =
    Stream.concat(prevStates.stream(), Stream.of(currentState)).map(isFinalState);
return statesFinality.filter(identity).count() == 1 && Iterables.getLast(statesFinality) == true;
```

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+
+  public class MockGaaSObservabilityProducer extends GaaSObservabilityEventProducer {
+    public MockGaaSObservabilityProducer(State state, MultiContextIssueRepository issueRepository) {
+      super(state, issueRepository);
+    }
+    // Send the events to the class test queue, so tests should not run concurrently
+    @Override
+    protected void sendUnderlyingEvent(GaaSObservabilityEventExperimental event) {
+      emittedEvents.add(event);
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java:97]: unlike the other test, the queue is non-`static`... so does it support concurrent test exec?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 8]
```diff
@@ -136,6 +140,12 @@ public <V> void onRetry(Attempt<V> attempt) {
             }
           }
         }));
+
+    if (instrumentationEnabled && ConfigUtils.getBoolean(config, GaaSObservabilityEventProducer.GAAS_OBSERVABILITY_EVENT_ENABLED, false)) {
+      this.eventProducer = Optional.of(GaaSObservabilityEventProducer.getEventProducer(config,  jobIssueEventHandler.getIssueRepository()));
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:145]: slightly unconventional to construct within from a class name, rather than to make the event producer a ctor param.  using reflection internally makes the tests more challenging to write, as it appears you've found.  of course, there's a place for reflection--just at a higher level than this class.

I do see how you need the issue repo from the issue handler to create the observability event producer.  how about whereever that handler gets created and injected, that's where you'd inject an optional event producer?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 9]
```diff
@@ -288,6 +303,15 @@ private static void modifyStateIfRetryRequired(org.apache.gobblin.configuration.
     state.removeProp(TimingEvent.FlowEventConstants.DOES_CANCELED_FLOW_MERIT_RETRY);
   }
 
+  private static boolean isStateTransitionToFinal(org.apache.gobblin.configuration.State currentState, List<org.apache.gobblin.configuration.State> prevStates) {
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:306]: nit: impl seems more like `isNewStateTransitionToFinal` or `isFirstStateTransitionToFinal`

**Review by phet — COMMENTED** [review]
nice work, glad to see this... looks nearly there!

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 2 above)
Comment (homatthew) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:233]: I don't have a strong preference for one or the other as long as the team agrees on a logically consistent pattern. This topic has come up in my past roles, so I figured I'd ask out of curiosity. Good points about the API being more self descriptive as opposed to an annotation like `@Nullable` which is more hidden

The argument against `Optional` as as follows:
- WIthout optional, you can have 2 possible states `null` or actual reference
- With optional, you can have 3 possible states `null`, `Optional.empty()`, `Optional.present()`

It's reasonable to try your best not to pass null as a parameter. (e.g. Optional.ofNullable(...)). But it does add visual clutter from a readability perspective that doesn't add much value when method overloading would be totally sufficient. 

This is especially true when there is only one optional parameter. Here, we don't have a bunch of permutations of options that the method caller could use (Which is probably a code smell on its own). 

Some debates in case you want to go down the rabbit hole / get nerd sniped:
https://rules.sonarsource.com/java/RSPEC-3553
https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments
http://dolszewski.com/java/java-8-optional-use-cases/
https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555

https://docs.oracle.com/javase/10/docs/api/java/util/Optional.html
```
Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance.
```

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 6 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:312]: I think it's impossible for a state to flip flop because of the way how we perform `mergeState`, which follows an ordered execution status of 
```
  private static final List<ExecutionStatus> ORDERED_EXECUTION_STATUSES = ImmutableList
      .of(ExecutionStatus.COMPILED, ExecutionStatus.PENDING, ExecutionStatus.PENDING_RESUME, ExecutionStatus.PENDING_RETRY,
          ExecutionStatus.ORCHESTRATED, ExecutionStatus.RUNNING, ExecutionStatus.COMPLETE,
          ExecutionStatus.FAILED, ExecutionStatus.CANCELLED);
```
I think there is a risk if there is a COMPLETE event followed by a FAILED or CANCELLED event but I believe the first "final" event state we run into is the one we want users to accept.

**Comment by arjun4084346** [comment]
I read the description of the PR, can you please elaborate more what benefit GaaSObservabilityEvents has over a GobblinTrackingEvent with name="GaaSObservabilityEvent". How is it more easily queryable compared to GobblinTrackingEvent? Is the frequency of emission is the only benefit?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/troubleshooter/JobIssueEventHandler.java (hunk 3 above)
Comment (Will-Lo) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/troubleshooter/JobIssueEventHandler.java:54]: With your suggestion to move the class as parameter we no longer need this :)

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java (hunk 7 above)
Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java:97]: No longer needed, now tests should support concurrent test execs

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 10]
```diff
@@ -136,6 +140,12 @@ public <V> void onRetry(Attempt<V> attempt) {
             }
           }
         }));
+
+    if (instrumentationEnabled && ConfigUtils.getBoolean(config, GaaSObservabilityEventProducer.GAAS_OBSERVABILITY_EVENT_ENABLED, false)) {
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:144]: Yes I think I'm going to do a better method of the determining whether to emit the event or not through a `NoopGaaSObservabilityEventSubmitter`, which is the default class and will do no work.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 2 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:233]: I think I'm going to lean towards going neither down the route of nulls or Optionals, and make a NoopEventSubmitter which will act as the default class from the factory, and its behavior is to do no work.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    return builder.build();
+  }
+
+  private static JobStatus convertExecutionStatusTojobState(State state, ExecutionStatus executionStatus) {
+    switch (executionStatus) {
+      case FAILED:
+        // TODO: Separate failure cases to SUBMISSION FAILURE and COMPILATION FAILURE, investigate events to populate these fields
+        if (state.contains(TimingEvent.JOB_END_TIME)) {
+          return JobStatus.EXECUTION_FAILURE;
+        }
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java:128]: It could be a submission failure currently, so it's hard to differentiate. I'm going to put that logic in a different PR I think to break this up

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+   * @return GaaSObservabilityEvent
+   */
+  private GaaSObservabilityEventExperimental createGaaSObservabilityEvent(State jobState) {
+    Long jobStartTime = jobState.contains(TimingEvent.JOB_START_TIME) ? jobState.getPropAsLong(TimingEvent.JOB_START_TIME) : null;
+    Long jobEndTime = jobState.contains(TimingEvent.JOB_END_TIME) ? jobState.getPropAsLong(TimingEvent.JOB_END_TIME) : null;
+    GaaSObservabilityEventExperimental.Builder builder = GaaSObservabilityEventExperimental.newBuilder();
+    List<Issue> issueList = null;
+    try {
+      issueList = issueRepository.getAll(TroubleshooterUtils.getContextIdForJob(jobState.getProperties())).stream().map(
+              issue -> new org.apache.gobblin.metrics.Issue(issue.getTime().toEpochSecond(),
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java:95]: Ah I think I was doing a conversion and had both classes, no longer needed though good catch

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java:128]: But to prevent a null I will mark as submission failure

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaAvroJobStatusMonitor.java (hunk 4 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaAvroJobStatusMonitor.java:68]: I'm going to leave instrumentationEnabled only for incrementing the metrics of the meters/gauges, and the whether or not the class is configured to emit events will be handled separately.

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java [hunk 13]
```diff
@@ -358,6 +370,66 @@ public void testProcessingRetriedForApparentlyTransientErrors() throws IOExcepti
     jobStatusMonitor.shutDown();
   }
 
+  @Test (dependsOnMethods = "testProcessingRetriedForApparentlyTransientErrors")
+  public void testProcessMessageForCancelledAndKilledEvent() throws IOException, ReflectiveOperationException {
```
Comment (phet) [inline gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:374]: sorry, can't recall: was this in the prior revision or just introduced?

I'm not clear whether it belongs in this PR or is unrelated...

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java [hunk 14]
```diff
@@ -387,62 +460,95 @@ public void testProcessProgressingMessageWhenNoPreviousStatus() throws IOExcepti
...
+    }
+    MultiContextIssueRepository issueRepository = new InMemoryMultiContextIssueRepository();
+    MockGaaSObservabilityEventProducer mockEventProducer = new MockGaaSObservabilityEventProducer(
+        ConfigUtils.configToState(ConfigFactory.empty()), issueRepository);
+    MockKafkaAvroJobStatusMonitor jobStatusMonitor = createMockKafkaAvroJobStatusMonitor(new AtomicBoolean(false), ConfigFactory.empty(),
+        mockEventProducer);
+    jobStatusMonitor.buildMetricsContextAndMetrics();
+    Iterator<DecodeableKafkaRecord> recordIterator = Iterators.transform(
+        this.kafkaTestHelper.getIteratorForTopic(TOPIC),
+        this::convertMessageAndMetadataToDecodableKafkaRecord);
```
Comment (phet) [inline gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:533]: perhaps not now... but definitely if you need to add further tests (as you fill in details of the observability event) - let's explore whether this repetitive boilerplate could live behind a reusable abstraction.

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java [hunk 15]
```diff
@@ -527,13 +637,14 @@ private GobblinTrackingEvent createGTE(String eventName, Map<String, String> cus
     return new GobblinTrackingEvent(timestamp, namespace, eventName, metadata);
   }
 
-  MockKafkaAvroJobStatusMonitor createMockKafkaAvroJobStatusMonitor(AtomicBoolean shouldThrowFakeExceptionInParseJobStatusToggle, Config additionalConfig) throws IOException, ReflectiveOperationException {
+  MockKafkaAvroJobStatusMonitor createMockKafkaAvroJobStatusMonitor(AtomicBoolean shouldThrowFakeExceptionInParseJobStatusToggle, Config additionalConfig,
+      GaaSObservabilityEventProducer eventProducer) throws IOException, ReflectiveOperationException {
```
Comment (phet) [inline gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:641]: your choice, but rather than updating so many calls to now take a no-op version, I might overload this method with a two-param form, that merely forwards to the three-param one here, after constructing the no-op instance.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    this.issueRepository = issueRepository;
+    this.instrumentationEnabled = instrumentationEnabled;
+    if (this.instrumentationEnabled) {
+      this.metricContext = Instrumented.getMetricContext(state, getClass());
+      this.getIssuesFailedMeter = this.metricContext.contextAwareMeter(MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX,
+          ISSUES_READ_FAILED_METRIC_NAME));
+    }
+  }
+
+  public void emitObservabilityEvent(State jobState) {
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java:74]: an argument cold be made for this to be `final`

update: below, I found out why you don't... perhaps document for maintainers?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 6 above)
Comment (phet) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:307]: agreed... that's what I meant by perhaps not possible.  so the difference would be purely stylistic... either form looks fine

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 5 above)
Comment (phet) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:302]: still thinking the same... why initialize on every invocation, rather than once when the class is loaded?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/NoopGaaSObservabilityEventProducer.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.apache.gobblin.service.monitoring;
+
+import org.apache.gobblin.configuration.State;
+import org.apache.gobblin.metrics.GaaSObservabilityEventExperimental;
+import org.apache.gobblin.runtime.troubleshooter.MultiContextIssueRepository;
+
+
+public class NoopGaaSObservabilityEventProducer extends GaaSObservabilityEventProducer {
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/NoopGaaSObservabilityEventProducer.java:25]: good use of the Null Pattern!

any javadoc?

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    Assert.assertEquals(event.getExecutorUrl(), "hostName");
+    Assert.assertEquals(event.getIssues().size(), 1);
+  }
+
+  private Issue createTestIssue(String summary, String code, IssueSeverity severity) {
+    return Issue.builder().summary(summary).code(code).time(ZonedDateTime.now()).severity(severity).build();
+  }
+
+
+  public class MockGaaSObservabilityProducer extends GaaSObservabilityEventProducer {
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java]: didn't I see this impl above?  couldn't it be shared there (e.g. if you were to make this `public static`)?

(or even put it into its own file)

**Review by phet — APPROVED** [review]
looks very good.  the tests came out clearer and simpler by using the ctor param for initialization, rather than reflection.

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java (hunk 13 above)
Comment (Will-Lo) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:374]: I realized that the tests were out of order in their dependsOn after I added more tests so I fixed that in the file, sorry for the confusion

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 5 above)
Comment (ZihanLi58) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:302]: I think we have FlowStatusGenerator.FINISHED_STATUSES which is essentially the same list, maybe reuse that one?

**Review by ZihanLi58 — APPROVED** [review]
Other than kip's comments, LGTM.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java (hunk 16 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/GaaSObservabilityEventProducer.java:74]: I can still set it as final but it wouldn't stop me from mutating the jobState (not that it needs to be done though)

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java (hunk 14 above)
Comment (Will-Lo) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:533]: Yeah I think I will definitely need to cut down on the boilerplate for future tests

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java (hunk 18 above)
Comment (Will-Lo) [reply gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/GaaSObservabilityProducerTest.java]: Ah good catch I'll make it to its own file

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java (hunk 15 above)
Comment (Will-Lo) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:641]: Yeah I realized that after, since I already made the changes I think I'll leave this for now but I'll do that if there's any additional configuration needed in future tests for this file

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java (hunk 13 above)
Comment (phet) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/KafkaAvroJobStatusMonitorTest.java:373]: no worries, makes sense

**Review by phet — APPROVED** [review]
looks great!
